libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -0,0 +1,310 @@
+                          offsets.indices);
+  }
+
+  auto deduplicate(match_store& matches, int64_t& row_number,
+                   const table_slice& slice, const type& ty,
+                   const arrow::Array& elements) const
+    -> generator<table_slice> {
+    size_t begin{};
+    const auto now = steady_clock::now();
+    // Logic adapted from the `unique` operator
+    for (size_t row = 0; row < slice.rows(); ++row) {
+      auto element_view = value_at(ty, elements, static_cast<int64_t>(row));
+      auto match_it = matches.find(element_view);
+      if (match_it == matches.end()) {
+        std::tie(match_it, std::ignore)
+          = matches.emplace(materialize(element_view), match_type{});
+      }
+      auto& match = match_it->second;
+      // This value hasn't been matched within the timeout,
+      // reset match count to zero
+      if (now - match.last_time > cfg_.timeout) {
+        match.count = 0;
+      }
+      match.last_time = now;
+      // Same as above, but for the distance / row number
+      if (row_number - match.last_row_number > cfg_.distance) {
+        match.count = 0;
+      }
+      match.last_row_number = row_number;
+      // If we're over the --limit, skip the current row
+      if (match.count >= cfg_.limit) {
+        co_yield subslice(slice, begin, row);
+        begin = row + 1;
+      } else {
+        ++match.count;
+      }
+      ++row_number;
+    }
+    co_yield subslice(slice, begin, slice.rows());
+  }
+
+  auto cleanup_matches(match_store& matches, int64_t row_number) const -> void {
+    if (cfg_.distance == std::numeric_limits<int64_t>::max()
+        && cfg_.limit == std::numeric_limits<int64_t>::max()) {
+      return;
+    }
+    const auto now = steady_clock::now();
+    // Erase stale matches
+    std::erase_if(matches, [&](const auto& elem) -> bool {
+      auto& match = elem.second;
+      if (row_number - match.last_row_number > cfg_.distance) {
+        return true;
+      }
+      if (now - match.last_time > cfg_.timeout) {
+        return true;
+      }
+      return false;
+    });
+  }
+
+  auto cleanup_indices(index_cache& indices) const -> void {
+    if (indices.size() < 128) {
+      return;
+    }
+    // Extract all nodes from `indices`,
+    // sort by `last_use` (descending),
+    // insert back only the ones that have been used most recently
+    std::vector<index_cache::node_type> nodes{};
+    nodes.reserve(indices.size());
+    while (not indices.empty()) {
+      nodes.emplace_back(indices.extract(indices.begin()));
+    }
+    TENZIR_ASSERT(indices.empty());
+    std::partial_sort(nodes.begin(), nodes.begin() + 64, nodes.end(),
+                      [](const auto& a, const auto& b) {
+                        return a.mapped().last_use > b.mapped().last_use;
+                      });
+    nodes.resize(64);
+    for (auto& node : nodes) {
+      indices.insert(std::move(node));
+    }
+  }
+
+  configuration cfg_{};
+};
```

Comment:
How did you choose 64, 128, and 256 as magic numbers here? Intuition or from testing?

Reply:
No objective rationale here, just pretty round numbers. In practice, 64/128/256 different schemas getting run through `deduplicate` might not even be realistic, but I figured I wanted some sort of a limit to not needlessly hog memory for a cache that can be easily and relatively cheaply recomputed.

Reply:
Makes sense to me. Mind putting that in the code somewhere, because I'm sure someone will stumble over these numbers in the future?